### ISSUE TITLE
Work out how asset-manager should work in development when uploading to S3

### ISSUE THREAD

**Issue by chrisroos** [alphagov/asset-manager#155]
We'll need to consider how asset-manager will work in development once we're uploading to S3 by default. I think we'll need to do this before we do PR #154.

**Comment by floehopper** [comment]
I wonder whether it might be worth running a [fake S3 server][1] on the developer VM if not in development generally.

[1]: https://github.com/jubos/fake-s3

**Comment by floehopper** [comment]
I'm going to move this out of the ["Proxy assets requests via nginx" milestone][1] into the ["Decommission NFS" milestone][2], because I don't think we *need* to do anything until serving from S3 is the default. We might want to rename the "Decommission NFS" milestone to something like "Make proxying from S3 via Nginx the default" or possibly split it into two milestones.

@chrisroos, @chrislo: Please feel free to move this back into the "Proxy assets requests via nginx" milestone if you disagree.

[1]: https://github.com/alphagov/asset-manager/milestone/4
[2]: https://github.com/alphagov/asset-manager/milestone/5

**Comment by floehopper** [comment]
I've opened #267 to address this, although we can't merge it until we've addressed the issue I mention in the PR description.

**Comment by floehopper** [comment]
I've just merged #267, so this should all be working, but I'd like to get someone else to try it out before I close this issue, so I'm going to re-open it for now.

**Comment by floehopper** [comment]
@chrisroos, @chrislo Would you be able to try the following on your development VM?

1. Set `PROXY_PERCENTAGE_OF_ASSET_REQUESTS_TO_S3_VIA_NGINX` to `100`
2. Set `PROXY_PERCENTAGE_OF_WHITEHALL_ASSET_REQUESTS_TO_S3_VIA_NGINX` to `100`
3. Ensure `AWS_S3_BUCKET_NAME` is not set (and ideally none of the AWS credential env vars are set)
4. Create and request a Mainstream asset using [the instructions in the README](https://github.com/alphagov/asset-manager#api-examples-development-vm)
5. Ensure the `GOVUK-Asset-Manager-File-Store` response header is set to `S3`
6. Create and request a Mainstream asset using [the instructions in the README](https://github.com/alphagov/asset-manager#create-a-whitehall-asset)
7. Ensure the `GOVUK-Asset-Manager-File-Store` response header is set to `S3`

**Comment by chrisroos** [comment]
I've just tried this, @floehopper, and it all works OK.

In testing the functionality, I realised that we're not yet creating the `PROXY_PERCENTAGE_OF_WHITEHALL_ASSET_REQUESTS_TO_S3_VIA_NGINX` environment variable in govuk-puppet. I'll open a pull request to do that now.

**Comment by chrisroos** [comment]
I've opened https://github.com/alphagov/govuk-puppet/pull/6710 to create the `PROXY_PERCENTAGE_OF_WHITEHALL_ASSET_REQUESTS_TO_S3_VIA_NGINX` environment variable.

**Comment by floehopper** [comment]
@chrisroos:

> I've just tried this, @floehopper, and it all works OK.

Great - thanks!

On the basis of @chrisroos' testing, I'm going to close this issue.

**Comment by floehopper** [comment]
@chrisroos:

> I've opened https://github.com/alphagov/govuk-puppet/pull/6710 to create the `PROXY_PERCENTAGE_OF_WHITEHALL_ASSET_REQUESTS_TO_S3_VIA_NGINX` environment variable.

I assumed we didn't need to create the env var until we needed to set it to anything other than zero. Was my assumption incorrect? Thanks for creating the PR in any case - we definitely will need the env var!

**Comment by chrisroos** [comment]
> I assumed we didn't need to create the env var until we needed to set it to anything other than zero. Was my assumption incorrect? Thanks for creating the PR in any case - we definitely will need the env var!

Your assumption was correct, @floehopper. I figured I might as well do it given that it didn't take too long and it made it easier to test the change in #267.
